src/internal/collection/postgres_collection.go

```diff
@@ -96,61 +90,68 @@ func (c *postgresCollection) ReadWrite(tx *sqlx.Tx) PostgresReadWriteCollection
 // NewSQLTx starts a transaction on the given DB, passes it to the callback, and
 // finishes the transaction afterwards. If the callback was successful, the
 // transaction is committed. If any errors occur, the transaction is rolled
-// back.  This will reattempt the transaction up to three times.
+// back.  This will reattempt the transaction forever.
 func NewSQLTx(ctx context.Context, db *sqlx.DB, apply func(*sqlx.Tx) error) error {
```

Comment:
We have something like this in `dbutil`.

Reply:
The implementations are different enough that we can't just use the `dbutil` one - `NewSQLTx`:
 * recognizes a custom `TransactionConflictErr` type as a reattempt indicator (used by `CreatePipelineInTransaction` to retry once the fileset is created)
 * recognizes all postgres transaction rollback errors instead of just serialization errors
 * doesn't do backoff for reattempts

In additon, we also have `NewDryrunSQLTx` which is different from doing `ReadOnly=true` because we actually do want writes to go through, just that they will be rolled back at the end (there may be better ways of doing this, but I suspect it will be complicated).

Reply:
It seems like these would all be improvements to the existing dbutil.WithTx, except for maybe the custom error. I don't think I have enough context to know why the custom error is needed.  Is it for a weaker isolation level?

Reply:
No, it's that `CreatePipelineInTransaction` has to do some reads before it can produce the new spec file, which it then writes out to a fileset.  The fileset is necessarily outside of the transaction, so it can't add that fileset to the spec commit unless we start the transaction over again.  So, really, the custom error is useful if the apply function needs to restart the transaction to pull in the latest state.